cases/closeSessionOnRecreateApplication/functional/CloseSessionOnRecreateApplicationCest.php

```diff
@@ -0,0 +1,26 @@
+<?php
+
+namespace tests;
+
+final class CloseSessionOnRecreateApplicationCest
+{
+    public function sessionIsPreservedInBetweenRequestsWithinSingleTest(FunctionalTester $I)
+    {
+        $I->amOnPage(['site/writesession']);
+
+        $I->amOnPage(['site/readsession']);
+
+        $I->dontSee('session non started');
```

Comment:
Instead of testing the output, you could (should?) directly test the application state:
```php
$I->assertTrue(\Yii::$app->session->isActive);
// Other tests here.

```

Reply:
I would say the exact opposite, for this reasons:

1. This is a bug-proof test: for such cases, the test shall demonstrate and mimic the real use-case where the bug happened, to better understand where and why it happened
1. The session state is only a part of the issue, I've extended the comment explaining why the tests are written in such way: https://github.com/Codeception/yii2-tests/pull/11/files#diff-87c8d3417e6d58d41f0bde02a29e6860R22-R26
1. This test use `recreateApplication: true`, a very deep flag to me: maintainers somewhere in the future could decide to drop the whole application state just after the request, and not only before the next request. I don't know if in the future the application state will still be readable in between requests, but these very test should not be affected by such possible change
1. Black box test are much more resilient and useful, they tell "I don't care how you do it, just get me what I want"

Reply:
Fair enough.